src/gpu/ForwardProjectorByBinNiftyPET.cxx

```diff
@@ -0,0 +1,190 @@
+    check(*this->_proj_data_info_sptr, *_density_sptr);
+    ForwardProjectorByBin::set_up(proj_data_info_sptr, density_info_sptr);
+    _symmetries_sptr.reset(new DataSymmetriesForBins_PET_CartesianGrid(proj_data_info_sptr, density_info_sptr));
+
+    // Initialise projected_data_sptr from this->_proj_data_info_sptr
+    _projected_data_sptr.reset(
+                new ProjDataInMemory(this->_density_sptr->get_exam_info_sptr(), this->_proj_data_info_sptr));
+}
+
+const DataSymmetriesForViewSegmentNumbers *
+ForwardProjectorByBinNiftyPET::
+get_symmetries_used() const
+{
+  if (!this->_already_set_up)
+    error("ForwardProjectorByBin method called without calling set_up first.");
+  return _symmetries_sptr.get();
+}
+
+void
+ForwardProjectorByBinNiftyPET::
+actual_forward_project(RelatedViewgrams<float>&,
+      const DiscretisedDensity<3,float>&,
+        const int min_axial_pos_num, const int max_axial_pos_num,
+        const int min_tangential_pos_num, const int max_tangential_pos_num)
+{
+    throw std::runtime_error("Need to use set_input() if wanting to use ForwardProjectorByBinNiftyPET.");
+}
+
+void
+ForwardProjectorByBinNiftyPET::
+actual_forward_project(RelatedViewgrams<float>& viewgrams,
+        const int min_axial_pos_num, const int max_axial_pos_num,
+        const int min_tangential_pos_num, const int max_tangential_pos_num)
+{
+//    if (min_axial_pos_num != _proj_data_info_sptr->get_min_axial_pos_num() ||
+//         … )
+//       error();
+
+    viewgrams = _projected_data_sptr->get_related_viewgrams(
+        viewgrams.get_basic_view_segment_num(), _symmetries_sptr);
+}
+
+void
+ForwardProjectorByBinNiftyPET::
+set_input(const shared_ptr<DiscretisedDensity<3,float> >& density_sptr)
+{
+    _density_sptr.reset(density_sptr->clone());
+
+    // Get dimensions of sinogram
+    int num_sinograms = _projected_data_sptr->get_num_sinograms();
+    int num_views     = _projected_data_sptr->get_num_views();
+    int num_tang_poss = _projected_data_sptr->get_num_tangential_poss();
+    int num_proj_data_elems = num_sinograms * num_views * num_tang_poss;
+    // Create array for sinogram and fill it
+    float *proj_data_ptr = new float[num_proj_data_elems];
+    // Necessary?
+    for (int i=0; i<num_proj_data_elems; ++i)
+        proj_data_ptr[i] = 0.F;
+
+    // Get dimensions of image
+    int dim[3];
+    Coordinate3D<int> min_indices;
+    Coordinate3D<int> max_indices;
+    if (!_density_sptr->get_regular_range(min_indices, max_indices))
+        throw std::runtime_error("ForwardProjectorByBinNiftyPET::set_input - "
+                                 "expected image to have regular range.");
+    for (int i=0; i<3; ++i)
+        dim[i] = max_indices[i + 1] - min_indices[i + 1] + 1;
+    // Create array for image and fill it - think carefully about index order!
+    float *im_ptr = new float[dim[0]*dim[1]*dim[2]];
+    // TODO: assert((y,x,z) = (320,320,128)) as required by NiftyPET by cropping/padding original 344x344x127
```

Comment:
NiftyPET's input image (im_ptr) must be cropped/padded to (z,x,y) = (128,320,320). Pretty sure Python's `[y, x, z]` becomes `[z, x, y]` in C++ (@pjmark?)